### DISCUSSION TITLE
Database object is getting resolved during startup in Elsa API Service

### DISCUSSION THREAD

**Discussion by vnmnm** [elsa-workflows/elsa-core#4063] [Q&A]
Hi,
I implemented a custom persistence provider and tried to test it by hosting the Elsa Server within a ASP.NET Web API project. Since I want my API to be multi-tenant and the connection string of the database gets resolved when the Http Request arrives depending on the `TenantId` claim in the authorization token in the request. However, after integrating the persistence provider, I am encountering an issue where my database object is getting resolved during application startup and since during startup there won't be any tenant context and hence no valid connection string, it fails to resolve that object.

I followed the steps outlined in this [link](https://elsa-workflows.github.io/elsa-core/docs/next/quickstarts/quickstarts-aspnetcore-server-api-endpoints) and below are the relevant snippets of code from my application.

Startup.cs
```
var elsaSection = Configuration.GetSection("Elsa");
services
    .AddElsa(elsa => elsa
        .UseObjectDBPersistence()
        .AddConsoleActivities()
        .AddHttpActivities(elsaSection.GetSection("Server").Bind)
        .AddWorkflowsFrom<Startup>()
    );

services.AddElsaApiEndpoints();
```
ObjectDB is our custom database provider which is just a wrapper around multiple databases that a tenant can choose to use.

```
public static ElsaOptionsBuilder UseObjectDBPersistence(this ElsaOptionsBuilder elsa)
{
    elsa.Services
        .AddScoped<ObjectDBWorkflowDefinitionStore>()
        .AddScoped<ObjectDBWorkflowInstanceStore>()
        .AddScoped<ObjectDBWorkflowExecutionLogStore>()
        .AddScoped<ObjectDBBookmarkStore>()
        .AddScoped<ObjectDBTriggerStore>();

    return elsa
        .UseWorkflowDefinitionStore(sp =>
        {
            return sp.GetRequiredService<ObjectDBWorkflowDefinitionStore>();
        })
        .UseWorkflowInstanceStore(sp =>
        {
            return sp.GetRequiredService<ObjectDBWorkflowInstanceStore>();
        })
        .UseWorkflowExecutionLogStore(sp =>
        {
            return sp.GetRequiredService<ObjectDBWorkflowExecutionLogStore>();
        })
        .UseBookmarkStore(sp =>
        {
            return sp.GetRequiredService<ObjectDBBookmarkStore>();
        })
        .UseTriggerStore(sp =>
        {
            return sp.GetRequiredService<ObjectDBTriggerStore>();
        });
}
```
When I run my application with these changes, it tries to resolve `ObjectDBWorkflowDefinitionStore` and other classes implemented for my custom persistence provider which will eventually fail as there is no tenant context. Below is the error which confirms it's trying to resolve those dependencies during startup. Attaching the exception details of the same. _I just commented my code related to my database registration as with those registrations it fails further down when resolving context which hides the error related to the actual object it was trying to resolve._

In Elsa code I saw that it registers some hosted services which might be causing these resolutions but in my app I am not adding any of those registrations that start any hosted service.

Not sure what I am missing w.r.t multi-tenancy. Any help would be greatly appreciated.

[exception-details.txt](https://github.com/elsa-workflows/elsa-core/files/11553425/exception-details.txt)

**Comment by sfmskywalker** [comment]
Multi tenancy isn't a fully-developed feature unfortunately. However, I am planning on adding proper support for multi-tenant applications in Elsa 3. Then perhaps, any findings I do there, can be applied to Elsa 2 as well.

**Reply by vnmnm** [reply]
Thanks you @sfmskywalker for the details. I will try similar approach with Elsa 3.0 latest pre-release packages.

**Reply by vnmnm** [reply]
@sfmskywalker I tried this with Elsa 3.0 RC packages and below are the details of the issue I am encountering. Relates more to FastEndpoints issue.

I copied an endpoint from one of the existing samples within the Elsa samples and injected an abstraction that resolves per tenant per request.

```
public class SlowEndpoint : EndpointWithoutRequest
{
    private readonly IRuntimeContext runtimeContext;

    public SlowEndpoint(IRuntimeContext runtimeContext)
    {
        this.runtimeContext = runtimeContext;
    }

    public override void Configure()
    {
        Get("/slow");
        AllowAnonymous();
    }

    public override async Task HandleAsync(CancellationToken ct)
    {
        var userId = runtimeContext.UserId;
        await Task.Delay(TimeSpan.FromSeconds(10), ct);
        await SendOkAsync("Slow endoint done");
    }
}
```
In my case, the `IRuntimeContext` is resolved per request and contains tenant related information. Once this sets, all other abstractions resolve as per this and connect to right tenant database etc. During startup, however, the endpoints are getting resolved and it's leading to error resolving the `IRuntimeContext` as it needs a valid bearer token in the request from which it read various claims to set the `TenantId` etc.

From call stack, I see that the Endpoint factory is creating the endpoints during startup which in turn is leading to resolution of `IRuntimeContext` which fails because the it uses a `DefaultHttpContext` during startup which will not have required bearer token in the request. Please see image snapshot below.

![image](https://github.com/elsa-workflows/elsa-core/assets/32916067/ebdf36a6-acd9-41e5-899f-63362683d25d)

From this issue on the FastEndpoints GitHub repo, it's mentioned that such constructor DI is not supported and property injection should be used instead. Please refer https://github.com/FastEndpoints/FastEndpoints/issues/60.

If I see any endpoint implemented by Elsa, it almost always injects some dependency that would need to connect to persistent store. When we implement our custom persistent store, then it would first try to resolve `IRuntimeContext` and then resolve `IWorkflowDefinitionStore` which will fail as seen above with `SlowEndpoint`.

```
internal class Get : ElsaEndpoint<Request>
{
    private readonly IWorkflowDefinitionStore _store;
    private readonly IApiSerializer _apiSerializer;
    private readonly WorkflowDefinitionMapper _mapper;

    public Get(IWorkflowDefinitionStore store, IApiSerializer apiSerializer, WorkflowDefinitionMapper mapper)
    {
        _store = store;
        _apiSerializer = apiSerializer;
        _mapper = mapper;
    }

    public override void Configure()
    {
        Get("/workflow-definitions/{definitionId}");
        ConfigurePermissions("read:workflow-definitions");
    }
}
```

So with this behavior of FastEndpoints, how  do we achieve multi-tenancy with Elsa 3.0? With controllers, this behavior won't be seen as they are not created during startup.

**Reply by vnmnm** [reply]
@sfmskywalker We also tried to implement our own custom persistent provider by following the changes at https://github.com/elsa-workflows/elsa-core/tree/mongodb/src/modules/Elsa.MongoDB and are running into similar errors due to Hosted Services which are trying to resolve the persistency related abstractions during startup as seen from below screenshot. There's a hosted service called `ResgiterDescriptors` which results in initialization of `IWorkflowDefinitionStore` which fails in case of a multi-tenant persistent store as there won't be a valid tenant context during startup.

![image](https://github.com/elsa-workflows/elsa-core/assets/32916067/14d2e7ed-0904-44c1-8b4b-933ddbda9dae)

**Reply by vnmnm** [reply]
@sfmskywalker I posted a question on the FastEndpoints repo and got a response that during startup we can check for null HttpContext and workaround this as the HttpContext will only ever be null during startup and this will not affect any subsequent requests. Please refer https://github.com/FastEndpoints/FastEndpoints/issues/455. This is working in my case. So please ignore the question related to slow endpoint posted above in this chain.

However, in case of Elsa, when we apply this workaround, I am not sure how it would impact as during startup the hosted services (shared above) that run do not have appropriate context which will resolve to a correct database and still fail.

**Reply by sfmskywalker** [reply]
Thanks a lot for the research and sharing your findings @vnmnm !
Indeed, the hosted services + multi-tenancy is a bit of a question mark at the moment, at least for me.

One idea to explore is to consider a base Hosted Service that basically loops over all tenant registrations, and then invoking some virtual `OnStartAsync` on the sub class with the current tenant.
